src/compiler.ts

```diff
@@ -6512,14 +6600,56 @@ export class Compiler extends DiagnosticEmitter {
         return module.unreachable();
       }
     }
-    return this.compileCallIndirect(
-      assert(signature), // FIXME: asc can't see this yet
-      indexArg,
-      expression.arguments,
-      expression,
-      0,
-      contextualType == Type.void
-    );
+
+    var returnType = signature.returnType;
+    var tempResultLocal = this.currentFlow.getTempLocal(returnType);
+    var tempFunctionReferenceLocal = this.currentFlow.getTempLocal(returnType);
+
+    var callExpr = module.block(null, [
+      module.local_set(
+        tempFunctionReferenceLocal.index,
+        indexArg
+      ),
+      module.push(
+        module.global_get(BuiltinNames.global_closure, NativeType.I32)
+      ),
+      module.local_set(
+        tempResultLocal.index, 
+        this.compileCallIndirect(
+          assert(signature), // FIXME: asc can't see this yet
+          this.ifClosure(
+            module.local_get(tempFunctionReferenceLocal.index, NativeType.I32),
+            module.block(null, [
+              module.global_set(
+                BuiltinNames.global_closure,
+                this.getClosurePtr(
+                  module.local_get(tempFunctionReferenceLocal.index, NativeType.I32),
+                )
+              ),
+              module.load(
+                4,
+                true,
+                module.global_get(BuiltinNames.global_closure, NativeType.I32),
+                NativeType.I32,
+                0
+              )
+            ], this.options.nativeSizeType),
+            module.local_get(tempFunctionReferenceLocal.index, NativeType.I32),
+          ),
+          expression.arguments,
+          expression,
+          0,
+          contextualType == Type.void
+        )
+      ),
+      module.global_set(BuiltinNames.global_closure, module.pop(NativeType.I32)),
+      module.local_get(tempResultLocal.index, returnType.toNativeType())
+    ], returnType.toNativeType())
+
+    this.currentFlow.freeTempLocal(tempResultLocal);
+    this.currentFlow.freeTempLocal(tempFunctionReferenceLocal);
```

Comment:
As I remember `freeTempLocal` should have LIFO order according it initialization. cc @dcodeIO